lib/puppet/resource/type.rb

```diff
@@ -105,6 +105,42 @@ def child_of?(klass)
     return(klass == parent_type ? true : parent_type.child_of?(klass))
   end
 
+  # Evaluate the resources produced by the given resource. These resources are
+  # evaluated in a separate but identical scope from the rest of the resource.
+  def evaluate_produces(resource, scope)
+    # Only defined types can produce capabilities
+    return unless definition?
+
+    resource.export.map do |ex|
+      blueprint = produces.find { |pr| pr[:capability] == ex.type } or
+        raise Puppet::ParseError, "Resource type #{resource.type} does not produce a #{ex.type}"
+      produced_resource = Puppet::Parser::Resource.new(ex.type, ex.title, :scope => scope, :source => self)
+
+      produced_resource.resource_type.parameters.each do |name|
+        next if name == :name
+
+        if expr = blueprint[:mappings][name.to_s]
+          produced_resource[name] = expr.safeevaluate(scope)
+        else
+          produced_resource[name] = scope[name.to_s]
+        end
+      end
+      # Tag the produced resource so we can later distinguish it from
+      # copies of the resource that wind up in the catalogs of nodes that
+      # use this resource. We tag the resource with producer:<environment>,
+      # meaning produced resources need to be unique within their
+      # environment
+      # @todo lutter 2014-11-13: we would really like to use a dedicated
+      # metadata field to indicate the producer of a resource, but that
+      # requires changes to PuppetDB and its API; so for now, we just use
+      # tagging
+      produced_resource.tag("producer:#{scope.catalog.environment}")
```

Comment:
I am somewhat concerned about the use of ':' in the tag. I know we do split tags on '::'. In general the spec of allowed characters in tags is weak (understatement), and it there may be parts of the ecosystem that could have difficulties - but it may also just work fine...